service/history/queues/iterator.go

```diff
@@ -0,0 +1,145 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"fmt"
+
+	"go.temporal.io/server/common/collection"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+type (
+	Iterator interface {
+		collection.Iterator[tasks.Task]
+
+		Range() tasks.Range
+		CanSplit(tasks.Key) bool
+		Split(key tasks.Key) (left Iterator, right Iterator)
+		CanMerge(Iterator) bool
+		Merge(Iterator) Iterator
+		Remaining() Iterator
+	}
+
+	paginationFnProvider func(tasks.Range) collection.PaginationFn[tasks.Task]
+
+	IteratorImpl struct {
+		paginationFnProvider paginationFnProvider
+		remainingRange       tasks.Range
+
+		pagingIterator collection.Iterator[tasks.Task]
+	}
+)
+
+func NewIterator(
+	paginationFnProvider paginationFnProvider,
+	r tasks.Range,
+) *IteratorImpl {
+	return &IteratorImpl{
+		paginationFnProvider: paginationFnProvider,
+		remainingRange:       r,
+
+		// lazy initialized to prevent task pre-fetching on creating the iterator
+		pagingIterator: nil,
+	}
+}
+
+func (i *IteratorImpl) HasNext() bool {
+	if i.pagingIterator == nil {
+		i.pagingIterator = collection.NewPagingIterator(i.paginationFnProvider(i.remainingRange))
+	}
+
+	return i.pagingIterator.HasNext()
+}
+
+func (i *IteratorImpl) Next() (tasks.Task, error) {
+	if !i.HasNext() {
+		panic("Iterator encountered Next call when there is no next item")
+	}
+
+	task, err := i.pagingIterator.Next()
+	if err != nil {
+		return nil, err
+	}
+
+	taskKey := task.GetKey()
+	taskID := taskKey.TaskID
+	if taskKey.FireTime.UnixNano() <= 0 {
+		// only increase taskID for immediate task category
+		taskID++
+	}
```

Comment:
why not define Next() on task key and call next here?

why only handle transfer task IDs? why timer task ID is not "++"ed?

Reply:
Synced offline, will always increase taskID and overflow to fire time if taskID is maxInt64. 

When querying db though, taskID used should be 0 for scheduled tasks (due to sanity check in persistence layer). This logic will be inside the function returned by `paginationFnProvider` so technically unrelated to iterator itself.